makefiles/info-global.inc.mk

```diff
@@ -1,6 +1,16 @@
 .PHONY: info-buildsizes info-buildsizes-diff info-features-missing \
         info-boards-features-missing
 
+# Perform dependency resolution without having included
+# $(RIOTBASE)/Makefile.features for now. This results in no optional modules and
+# no architecture specific handling being done. The result will be a subset of
+# used modules and requirements that is present in for *all* boards, so this can
+# be cached to speed up individual dependency checks
+include $(RIOTMAKE)/defaultmodules.inc.mk
+# add default modules
+USEMODULE += $(filter-out $(DISABLE_MODULE),$(DEFAULT_MODULE))
+include $(RIOTMAKE)/dependency_resolution.inc.mk
```

Comment:
I realize that this will mean that the output of `info-boards-supported` will not match the actual dependencies for applications disabling modules in their `Makefile`, or `BOARDS` declaring `DEFAULT_MODULE`. Any dependency of those modules will never be included.

I don't think this currently causes any issue (in the sense that the output is not false), but it could eventually lead to different resolution with `info-boards-supported` and actually trying to compile the configuration on one of the supposedly `supported-boards`.

Reply:
> I realize that this will mean that the output of `info-boards-supported` will not match the actual dependencies for applications disabling modules in their `Makefile`, or `BOARDS` declaring `DEFAULT_MODULE`. Any dependency of those modules will never be included.

This is not true. The static dependency resolution done there is only as used as "seed". This allows speeding up the dependency resolution of the individual boards, in which indeed all the board specific stuff is handled. This "seed" only reduces the number of recursions needed before the list of modules and packages used converges.

Also: This seed is also checked for each board before doing the dependency resolution. If a board is not satisfying this "common set" of required modules and features, it is proven to not be compatible with the application. Only if the board is supporting the "seed" features, a full dependency resolution is done to make sure to not print boards that are supported.

(This is actually also what the comment in the code explains.)

Reply:
I might be missing something then, but currently we have: 

```
include $(RIOTMAKE)/defaultmodules.inc.mk
# add default modules
USEMODULE += $(filter-out $(DISABLE_MODULE),$(DEFAULT_MODULE))
```

and the a first round through `include $(RIOTMAKE)/dependency_resolution.inc.mk`

but during the first round `USEMODULE` and `DEFAULT_MODULE` have not been expanded, so during this first pass `USEMODULE` and `DEFAULT_MODULE` can be udpated, which also means that `DEFAULT_MODULE` added during that pass can be converted to `USEMODULE`.

But here since `USEMODULE` and `DEFAULT_MODULE` are expanded right away so no `DEFAULT_MODULE` is added into `USEMODULE` until the last iteration. So in the case of `stdio_%` if they where declared as a `DEFAULT_MODULE` they would not have become `USEMODULE` until the end and so there dependencies would not be added.

Maybe this is already taken into account somewhere else? And I'm missing this..

Reply:
I'm not sure if I understand the problem you are referring to.

> But here since `USEMODULE` and `DEFAULT_MODULE` are expanded right away

Where are they expended right away? Maybe I'm missing that to understand what you are referring to.

Reply:
So here is a very simplified representation of what happens in `makefile.include`

```
include $(RIOTMAKE)/defaultmodules.inc.mk
# add default modules

# recursive so DEFAULT_MODULE can be modified and through that modification modify USEMODULE
USEMODULE += $(filter-out $(DISABLE_MODULE),$(DEFAULT_MODULE))

include $(RIOTMAKE)/dependency_resolution.inc.mk

# This actually happens inside dependency_resolution.inc.mk at the end of every iteration
USEMODULE := $(sort $(USEMODULE))
 
... iterations...

# Add those modules that where not added in the first iteration
USEMODULE += $(filter-out $(DISABLE_MODULE),$(DEFAULT_MODULE))
```

What happens in `info-global.inc.mk`:

```
include $(RIOTMAKE)/defaultmodules.inc.mk
# add default modules

USEMODULE += $(filter-out $(DISABLE_MODULE),$(DEFAULT_MODULE))

# simple expansion before going through the first iteration of dependency_resolution.inc.mk.
# No new DEFAULT module will be added until the last iteration
USEMODULE:= $(USEMODULE_GLOBAL)
DEFAULT_MODULE := $(DEFAULT_MODULE_GLOBAL)

include $(RIOTMAKE)/dependency_resolution.inc.mk

# This actually happens inside dependency_resolution.inc.mk  at the end of every iteration
USEMODULE := $(sort $(USEMODULE))

... iterations...

# Add all DEFAULT_MODULES that where not in defaultmodules.inc.mk
USEMODULE += $(filter-out $(DISABLE_MODULE),$(DEFAULT_MODULE))
```

So `USEMODULE` is simply expanded before going through `dependency_resolution.inc.mk`, where before it was simply-expanded after going though it a first time. This means that modules like `oneway-malloc` are added at the end of all the resolution instead of in the first iteration. Same for `stdio_%`.

Reply:
Note: I think `stdio_%` is the only module that has dependencies, maybe we can add special handling for this. And we could leave it out of scope of this PR. As I was saying in #13651 I would rather have this kind of behavior be not aloud.

`oweway_malloc` will go away with the msp430 toolchain change.

Reply:
> # simple expansion before going through the first iteration of dependency_resolution.inc.mk.
> # No new DEFAULT module will be added until the last iteration
> USEMODULE:= $(USEMODULE_GLOBAL)
> DEFAULT_MODULE := $(DEFAULT_MODULE_GLOBAL)

OK. Now I understand. But:
1. The behavior you are pointing out is the exact behavior currently in master
2. This PR actually improves on this, as static dependency resolution of the commonly required features and used modules has a recursive definition of `USEMODULE`. So for that resolution the behavior is indeed the same when doing it via `make info-boards-supported` and the "normal" dependency resolution. So now issues only arise if during the first recursion board/architecture specific stuff is added to `DEFAULT_MODULE` that in addition have dependencies to take into account.
3. This highlights that the current handling of `DEFAULT_MODULE` is complex. I'm glad that you opened https://github.com/RIOT-OS/RIOT/issues/13651 to iron out what the expected semantics and behavior of `DEFAULT_MODULE` is

Reply:
> The behavior you are pointing out is the exact behavior currently in master

True... I hadn't realized that... I guess every time I look at how `DEFAULT_MODULE` is handled I find more weird stufff..

So this is again oy of scope, thanks for pointing it out!

Reply:
Issue becomes irrelevant with #13785